### PR TITLE
Emit arrow function es6

### PR DESCRIPTION
**Pull request by yuit**
Emit Arrow function natively in ES6 and capture "this"

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -3210,6 +3210,13 @@ module ts {
                 emitTrailingComments(node);
             }
 
+            function isES6ArrowFunction(node: FunctionLikeDeclaration): boolean {
+                if (node.kind === SyntaxKind.ArrowFunction && compilerOptions.target >= ScriptTarget.ES6) {
```
Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: ```return node.kind === ... && compilerOptions.target ...;```

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -3258,6 +3271,12 @@ module ts {
                 tempVariables = undefined;
                 tempParameters = undefined;
                 emitSignatureParameters(node);
```
Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: it looks like if you have:

```var v = a => { };```

then we'll emit that as:

```var v = (a) => { };```

Can we make it so we emit this case idiomatically as it was originally written?

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -3269,7 +3288,12 @@ module ts {
                     startIndex = emitDirectivePrologues((<Block>node.body).statements, /*startWithNewLine*/ true);
                 }
                 var outPos = writer.getTextPos();
-                emitCaptureThisForNodeIfNecessary(node);
+
+                // In ES6, fat arrow function lexically binds this value. Therefore, when targeting es6, we can omit capturing of "this" in the fat arrow function
+                if (!isES6ArrowFunction(node)) {
+                    emitCaptureThisForNodeIfNecessary(node);
```
Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: why not have emitCaptureForThisNodeIfNecessary check if it's an ES6 arrow function or not?  I would think that would fall under hte bucket of 'if necessary'.

File: src/compiler/emitter.ts (hunk 2 above)
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Agreed

File: src/compiler/emitter.ts (hunk 3 above)
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Also, looks like in the checker, you already check for ES6 when setting needToCaptureLexicalThis

File: tests/baselines/reference/emitArrowFunction.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+//// [emitArrowFunction.ts]
```
Comment (JsonFreeman) [inline tests/baselines/reference/emitArrowFunction.js]: It would be really helpful if the baseline stated the target, just like the original test.

**Comment by DanielRosenwasser** [comment]
Are you still working off of this branch @yuit?

**Comment by yuit** [comment]
@DanielRosenwasser  Yes. Fixing up thing now. Almost done

**Comment by yuit** [comment]
Address Issue#1609. Also should we update typescript spec.?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -4836,6 +4836,16 @@ module ts {
         function checkIdentifier(node: Identifier): Type {
             var symbol = getResolvedSymbol(node);
 
+			// As noted in ECMAScript 6 language spec, arrow functions never have an arguments objects.
+			// Although in down-level emit of arrow function, we emit it using function expression which means that
+			// arguments objects will be inner bound while emitting arrow function natively in ES6, arguments objects
+			// will be bound to non-arrow function that contain this arrow function. This results in inconsistent bahaviour.
+			// To avoid that we will give an error to users if they use arguments objects in arrow function so that they
+			// can explicitly bound arguments objects
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Tabs.

File: src/compiler/diagnosticInformationMap.generated.ts [hunk 6]
```diff
@@ -452,5 +452,6 @@ module ts {
         You_cannot_rename_this_element: { code: 8000, category: DiagnosticCategory.Error, key: "You cannot rename this element." },
         yield_expressions_are_not_currently_supported: { code: 9000, category: DiagnosticCategory.Error, key: "'yield' expressions are not currently supported.", isEarly: true },
         Generators_are_not_currently_supported: { code: 9001, category: DiagnosticCategory.Error, key: "Generators are not currently supported.", isEarly: true },
+        An_argument_object_has_different_behaviour_across_Javascript_versions_Use_function_expression_or_rest_parameters_instead: { code: 9002, category: DiagnosticCategory.Error, key: "An argument object has different behaviour across Javascript versions. Use function expression or rest parameters instead" },
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticInformationMap.generated.ts]: Change this to

`The 'arguments' object is not allowed in an arrow function. Use a function expression or rest parameters instead.`

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -4890,7 +4900,14 @@ module ts {
...
             if (container.kind === SyntaxKind.ArrowFunction) {
                 container = getThisContainer(container, /* includeArrowFunctions */ false);
-                needToCaptureLexicalThis = true;
+
+                // When targeting es6, arrow function lexically bind "this" so we do not need to do the work of binding "this" in emitted code
+                if (compilerOptions.target >= ScriptTarget.ES6) {
+                    needToCaptureLexicalThis = false;
+                }
+				else {
+					needToCaptureLexicalThis = true;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Remove this branch, it's already set to `false`.

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -4890,7 +4900,14 @@ module ts {
             // Now skip arrow functions to get the "real" owner of 'this'.
             if (container.kind === SyntaxKind.ArrowFunction) {
                 container = getThisContainer(container, /* includeArrowFunctions */ false);
-                needToCaptureLexicalThis = true;
+
+                // When targeting es6, arrow function lexically bind "this" so we do not need to do the work of binding "this" in emitted code
+                if (compilerOptions.target >= ScriptTarget.ES6) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Reduce this to one branch, it's already set to `false`.

File: src/compiler/emitter.ts [hunk 9]
```diff
@@ -3263,6 +3263,10 @@ module ts {
                 emitTrailingComments(node);
             }
 
+            function isES6ArrowFunction(node: FunctionLikeDeclaration): boolean {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Just consolidate this into a local variable.

File: src/compiler/emitter.ts [hunk 10]
```diff
@@ -3272,7 +3276,13 @@ module ts {
                     // Methods will emit the comments as part of emitting method declaration
                     emitLeadingComments(node);
                 }
-                write("function ");
+
+                // For targeting below es6, emit functions-like declaration including arrow function using function keyword.
+                // When targeting ES6, emit arrow function natively in ES6 by omitting function keyword and using fat arrow instead
+                if (!isES6ArrowFunction(node)) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Use the local variable here.

File: src/compiler/emitter.ts [hunk 11]
```diff
@@ -3303,6 +3313,21 @@ module ts {
                 decreaseIndent();
             }
 
+			function emitSignatureParametersES6(node: FunctionLikeDeclaration) {
+				// Check the node's parameters whether it contains flags indicating that it has no parenthesis around the parameters
+				// Preserver no-parenthesis
+				if (node && node.flags & NodeFlags.SimpleArrowFunction) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Parens around the & expression please

File: src/compiler/emitter.ts [hunk 12]
```diff
@@ -3303,6 +3313,21 @@ module ts {
                 decreaseIndent();
             }
 
+			function emitSignatureParametersES6(node: FunctionLikeDeclaration) {
+				// Check the node's parameters whether it contains flags indicating that it has no parenthesis around the parameters
+				// Preserver no-parenthesis
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]:
```TypeScript
// Check whether the parameter list needs parentheses.
```

File: src/compiler/emitter.ts (hunk 11 above)
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Also why would `node` be `undefined`?

File: src/compiler/emitter.ts [hunk 13]
```diff
@@ -3303,6 +3313,21 @@ module ts {
                 decreaseIndent();
             }
 
+			function emitSignatureParametersES6(node: FunctionLikeDeclaration) {
+				// Check the node's parameters whether it contains flags indicating that it has no parenthesis around the parameters
+				// Preserver no-parenthesis
+				if (node && node.flags & NodeFlags.SimpleArrowFunction) {
+					increaseIndent();
+                    var parameters = node.parameters;
+                    var omitCount = languageVersion < ScriptTarget.ES6 && hasRestParameters(node) ? 1 : 0;
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: `languageVersion <= ScriptTarget.ES5` or `!(languageVersion >= ScriptTarget.ES6)`

File: src/compiler/types.ts [hunk 14]
```diff
@@ -293,6 +293,8 @@ module ts {
         Const =             0x00001000,  // Variable declaration
         OctalLiteral =      0x00002000,
 
+        SimpleArrowFunction =      0x00004000,  // Arrow function without parenthesized parameters
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: Do we really need this flag? Are there any syntactic ambiguities we can get into?

I'd say that if an arrow function
  * has a single parameter and
  * that parameter has no default initializer and
  * that parameter is not a rest parameter

then emit it as a simple arrow function.

Comment (JsonFreeman) [inline src/compiler/types.ts]: You could do the trick where you compare the beginning of the function expression to the beginning of the parameter. So if it has one parameter, whose pos coincides with the pos of the function expression, it's a simple arrow

File: src/compiler/emitter.ts [hunk 15]
```diff
@@ -3310,7 +3335,16 @@ module ts {
...
-                emitSignatureParameters(node);
+
+                // When targeting ES6, emit arrow function natively in ES6
+                if (isES6ArrowFunction(node)) {
+                    emitSignatureParametersForArrow(node);
+					write(" =>");
+                }
+				else {
+					emitSignatureParameters(node);
+				}
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Still tabs - just do a global project find for tabs in the `src` directory.

File: src/compiler/emitter.ts (hunk 11 above)
Comment (yuit) [inline src/compiler/emitter.ts]: Agree with you on that. I just check to make sure but you are right this shouldn't happen

File: src/compiler/types.ts (hunk 14 above)
Comment (yuit) [inline src/compiler/types.ts]: @DanielRosenwasser : Simple in this case is has no parenthesis. I don't think you can just check these three conditions. 
@JsonFreeman : That would work too

File: src/compiler/diagnosticInformationMap.generated.ts (hunk 6 above)
Comment (mhegazy) [inline src/compiler/diagnosticInformationMap.generated.ts]: +1

File: src/compiler/checker.ts (hunk 8 above)
Comment (mhegazy) [inline src/compiler/checker.ts]: needToCaptureLexicalThis = compilerOptions.target >= ScriptTarget.ES6

File: src/compiler/emitter.ts [hunk 16]
```diff
@@ -3303,6 +3313,20 @@ module ts {
                 decreaseIndent();
             }
 
+            function emitSignatureParametersForArrow(node: FunctionLikeDeclaration) {
+                // Check whether the parameter list needs parentheses and preserve no-parenthesis
+                if (node.flags & NodeFlags.SimpleArrowFunction) {
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: i do not think you need this flag.

**Comment by DanielRosenwasser** [comment]
`diagnosticMessages.json` is all tabs, in `master` it's all spaces, you might want to pull in from `master` and make sure there are no tabs.

**Comment by yuit** [comment]
@DanielRosenwasser  Yes there are tabs

**Comment by JsonFreeman** [comment]
:+1:

**Comment by yuit** [comment]
@basarat I have updated the error message in the last few commits

**Comment by basarat** [comment]
> I have updated the error message in the last few commits

sorry. my bad.

**Comment by yuit** [comment]
@basarat  no worry :) Thanks for code reviewing!

**Pull request merged**